lighthouse-core/audits/script-treemap-data.js

```diff
@@ -0,0 +1,262 @@
+const UnusedJavaScriptSummary = require('../computed/unused-javascript-summary.js');
+const ModuleDuplication = require('../computed/module-duplication.js');
+
+/**
+ * @typedef {RootNodeContainer[]} TreemapData
+ */
+
+/**
+ * Ex: https://gist.github.com/connorjclark/0ef1099ae994c075e36d65fecb4d26a7
+ * @typedef RootNodeContainer
+ * @property {string} name Arbitrary name identifier. Usually a script url.
+ * @property {Node} node
+ */
+
+/**
+ * @typedef Node
+ * @property {string} name Arbitrary name identifier. Usually a path component from a source map.
+ * @property {number} resourceBytes
+ * @property {number=} unusedBytes
+ * @property {string=} duplicatedNormalizedModuleName If present, this module is a duplicate. String is normalized source path. See ModuleDuplication.normalizeSource
+ * @property {Node[]=} children
+ */
+
+/**
+ * @typedef {Omit<Node, 'name'|'children'>} SourceData
+ */
+
+class ScriptTreemapDataAudit extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'script-treemap-data',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: 'Script Treemap Data',
+      description: 'Used for treemap app',
+      requiredArtifacts:
+        ['traces', 'devtoolsLogs', 'SourceMaps', 'ScriptElements', 'JsUsage', 'URL'],
+    };
+  }
+
+  /**
+   * Returns a tree data structure where leaf nodes are sources (ie. real files from source tree)
+   * from a source map, and non-leaf nodes are directories. Leaf nodes have data
+   * for bytes, coverage, etc., when available, and non-leaf nodes have the
+   * same data as the sum of all descendant leaf nodes.
+   * @param {string} sourceRoot
+   * @param {Record<string, SourceData>} sourcesData
+   * @return {Node}
+   */
+  static prepareTreemapNodes(sourceRoot, sourcesData) {
+    /**
+     * @param {string} name
+     * @return {Node}
+     */
+    function newNode(name) {
+      return {
+        name,
+        resourceBytes: 0,
+      };
+    }
+
+    const topNode = newNode(sourceRoot);
+
+    /**
+     * Given a slash-delimited path, traverse the Node structure and increment
+     * the data provided for each node in the chain. Creates nodes as needed.
+     * Ex: path/to/file.js will find or create "path" on `node`, increment the data fields,
+     *     and continue with "to", and so on.
+     * @param {string} source
+     * @param {SourceData} data
+     */
+    function addAllNodesInSourcePath(source, data) {
+      let node = topNode;
+
+      // Apply the data to the topNode.
+      topNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) topNode.unusedBytes = (topNode.unusedBytes || 0) + data.unusedBytes;
+
+      // Strip off the shared root.
+      const sourcePathSegments = source.replace(sourceRoot, '').split(/\/+/);
+      sourcePathSegments.forEach((sourcePathSegment, i) => {
+        const isLastSegment = i === sourcePathSegments.length - 1;
```

Comment:
nit: `isLastSegment` is a little confusing because it's in terms of path segments, while most of this code is thinking in terms of trees. Consider something like `isLeaf`?